hypothesis-python/docs/reference/targeted.rst

```diff
@@ -0,0 +1,23 @@
+Targeted property-based testing
+===============================
+
+Targeted property-based testing combines the advantages of both search-based and property-based testing.  Instead of being completely random, targeted PBT uses a search-based component to guide the input generation towards values that have a higher probability of falsifying a property.  This explores the input space more effectively and requires fewer tests to find a bug or achieve a high confidence in the system being tested than random PBT. (`Löscher and Sagonas <http://proper.softlab.ntua.gr/Publications.html>`__)
+
+This is not *always* a good idea - for example calculating the search metric might take time better spent running more uniformly-random test cases, or your target metric might accidentally lead Hypothesis *away* from bugs - but if there is a natural metric like "floating-point error", "load factor" or "queue length", we encourage you to experiment with targeted testing.
+
+.. code-block:: python
+
+  from hypothesis import given, strategies as st, target
+
+
+  @given(st.floats(0, 1e100), st.floats(0, 1e100), st.floats(0, 1e100))
+  def test_associativity_with_target(a, b, c):
+      ab_c = (a + b) + c
+      a_bc = a + (b + c)
+      difference = abs(ab_c - a_bc)
+      target(difference)  # Without this, the test almost always passes
```

Comment:
As fas as I can tell, `target` makes little difference here. Around 40% discover rate with or without it.

Reply:
same..I'm not going to follow up on this right now, but I've added it as a TODO to https://github.com/HypothesisWorks/hypothesis/issues/4309

Reply:
Thanks! A comment for that time (ignore for this PR) is that `target` likely doesn't work well here because the difference is constant for large regions of the input space.

We might want to comment that the target should be fairly smooth, not discrete-valued, and not added willy-nilly as it may actually reduce bug-finding power with the default example budget (as indicated below).

For background, the optimiser seems to get more stuck than I'd expect in these flat regions. E.g.,
```
@given(st.floats())
@settings(database=None, phases=[Phase.generate, Phase.target])
def test_target(f):
    target(f > 1e5)
    print(f)
```
doesn't vary `f` very much inside the `True` region. Fair enough, but worth noting.